File: python/langsmith/evaluation/llm_evaluator.py
"""Contains the LLMEvaluator class for building LLM-as-a-judge evaluators."""

from typing import Any, Callable, List, Optional, Tuple, Union

from pydantic import BaseModel

from langsmith.evaluation import EvaluationResult, EvaluationResults, RunEvaluator
from langsmith.schemas import Example, Run


class CategoricalScoreConfig(BaseModel):
    """Configuration for a categorical score."""

    key: str
    choices: List[str]
    description: str
    include_explanation: bool = False


class ContinuousScoreConfig(BaseModel):
    """Configuration for a continuous score."""

    key: str
    min: float = 0
    max: float = 1
    description: str
    include_explanation: bool = False


def _create_score_json_schema(
    score_config: Union[CategoricalScoreConfig, ContinuousScoreConfig],
) -> dict:
    properties: dict[str, Any] = {}
    if isinstance(score_config, CategoricalScoreConfig):
        properties["score"] = {
            "type": "string",
            "enum": score_config.choices,
            "description": f"The score for the evaluation, one of "
            f"{', '.join(score_config.choices)}.",
        }
    elif isinstance(score_config, ContinuousScoreConfig):
        properties["score"] = {
            "type": "number",
            "minimum": score_config.min,
            "maximum": score_config.max,
            "description": f"The score for the evaluation, between "
            f"{score_config.min} and {score_config.max}, inclusive.",
        }
    else:
        raise ValueError("Invalid score type. Must be 'categorical' or 'continuous'")

    if score_config.include_explanation:
        properties["explanation"] = {
            "type": "string",
            "description": "The explanation for the score.",
        }

    return {
        "title": score_config.key,
        "description": score_config.description,
        "type": "object",
        "properties": properties,
        "required": (
            ["score", "explanation"] if score_config.include_explanation else ["score"]
        ),
    }


class LLMEvaluator(RunEvaluator):
    """A class for building LLM-as-a-judge evaluators."""

    def __init__(
        self,
        *,
        prompt_template: Union[str, List[Tuple[str, str]]],
        score_config: Union[CategoricalScoreConfig, ContinuousScoreConfig],
        map_variables: Optional[Callable[[Run, Example], dict]] = None,
        model: Optional[str] = "gpt-3.5-turbo",
        model_provider: Optional[str] = "openai",
        **kwargs,
    ):
        """Initialize the LLMEvaluator.

        Args:
            prompt_template (Union[str, List[Tuple[str, str]]): The prompt
                template to use for the evaluation. If a string is provided, it is
                assumed to be a system message.
            score_config (Union[CategoricalScoreConfig, ContinuousScoreConfig]):
                The configuration for the score, either categorical or continuous.
            map_variables (Optional[Callable[[Run, Example], dict]], optional):
                A function that maps the run and example to the variables in the
            prompt. Defaults to None. If None, it is assumed that the prompt
                only requires 'input', 'output', and 'expected'.
            model (Optional[str], optional): The model to use for the evaluation.
                Defaults to "gpt-3.5-turbo".
            model_provider (Optional[str], optional): The model provider to use
                for the evaluation. Defaults to "openai".
        """
        try:
            from langchain_core.prompts import ChatPromptTemplate
        except ImportError as e:
            raise ImportError(
                "LLMEvaluator requires langchain-core to be installed. "
                "Please install langchain-core by running `pip install langchain-core`."
            ) from e
        try:
            from langchain.chat_models import init_chat_model
        except ImportError as e:
            raise ImportError(
                "LLMEvaluator requires langchain to be installed. "
                "Please install langchain by running `pip install langchain`."
            ) from e
        if isinstance(prompt_template, str):
            self.prompt = ChatPromptTemplate.from_messages(
                [("system", prompt_template)]
            )
        else:
            self.prompt = ChatPromptTemplate.from_messages(prompt_template)

        if set(self.prompt.input_variables) - {"input", "output", "expected"}:
            if not map_variables:
                raise ValueError(
                    "map_inputs must be provided if the prompt template contains "
                    "variables other than 'input', 'output', and 'expected'"
                )
        self.map_variables = map_variables

        self.score_config = score_config
        self.score_schema = _create_score_json_schema(self.score_config)

        try:
            model = init_chat_model(
                model=model, model_provider=model_provider, **kwargs
            ).with_structured_output(self.score_schema)
        except ImportError as e:
            raise ImportError(
                "LLMEvaluator is missing a required langchain integration."
            ) from e
        except ValueError as e:
            raise ValueError(
                "Error loading the model. Please check the model, model_provider, "
                "and that the appropriate secrets are set."
            ) from e

        self.runnable = self.prompt | model

    def evaluate_run(
        self, run: Run, example: Optional[Example] = None
    ) -> Union[EvaluationResult, EvaluationResults]:
        """Evaluate a run."""
        if self.map_variables:
            variables = self.map_variables(run, example)
            if set(self.prompt.input_variables) - set(variables.keys()):
                raise ValueError(
                    "map_variables must return a dictionary with keys for all of the "
                    "variables in the prompt. Expected variables: "
                    f"{self.prompt.input_variables}. Returned variables: "
                    f"{variables.keys()}"
                )
            output = self.runnable.invoke(variables)
        else:
            variables = {}
            if "input" in self.prompt.input_variables:
                if len(run.inputs) == 0:
                    raise ValueError(
                        "No input keys are present in run.inputs but the prompt "
                        "requires 'input'."
                    )
                if len(run.inputs) != 1:
                    raise ValueError(
                        "Multiple input keys are present in run.inputs. Please provide "
                        "a map_variables function."
                    )
                variables["input"] = list(run.inputs.values())[0]
            if "output" in self.prompt.input_variables:
                if len(run.outputs) == 0:
                    raise ValueError(
                        "No output keys are present in run.outputs but the prompt "
                        "requires 'output'."
                    )
                if len(run.outputs) != 1:
                    raise ValueError(
                        "Multiple output keys are present in run.outputs. Please "
                        "provide a map_variables function."
                    )
                variables["output"] = list(run.outputs.values())[0]
            if "expected" in self.prompt.input_variables:
                if not example:
                    raise ValueError(
                        "No example is provided but the prompt requires 'expected'."
                    )
                if len(example.outputs) == 0:
                    raise ValueError(
                        "No output keys are present in example.outputs but the prompt "
                        "requires 'expected'."
                    )
                if len(example.outputs) != 1:
                    raise ValueError(
                        "Multiple output keys are present in example.outputs. Please "
                        "provide a map_variables function."
                    )
                variables["expected"] = list(example.outputs.values())[0]
            output = self.runnable.invoke(variables)

        if isinstance(self.score_config, CategoricalScoreConfig):
            value = output["score"]
            explanation = output.get("explanation", None)
            return EvaluationResult(
                key=self.score_config.key, value=value, comment=explanation
            )
        elif isinstance(self.score_config, ContinuousScoreConfig):
            score = output["score"]
            explanation = output.get("explanation", None)
            return EvaluationResult(
                key=self.score_config.key, score=score, comment=explanation
            )

File: js/src/evaluation/llm_evaluator.ts
import { initChatModel } from "langchain/chat_models/universal";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import * as uuid from "uuid";
import {
  EvaluationResult,
  EvaluationResults,
  RunEvaluator,
} from "./evaluator.js";
import type { Run, Example } from "../schemas.js";

/**
 * Configuration for categorical (enum-based) scoring in evaluations.
 * Used to define discrete categories or labels for evaluation results.
 */
export class CategoricalScoreConfig {
  /** Feedback key for the evaluator */
  key: string;
  /** Array of valid categorical choices/labels that can be assigned */
  choices: string[];
  /** Description of what this score measures or represents */
  description: string;
  /** Optional key for the LLM reasoning/explanation for the score */
  reasoningKey?: string;
  /** Optional description of score reasoning, provided to the LLM in the structured output */
  reasoningDescription?: string;

  /**
   * Creates a new categorical score configuration
   * @param params Configuration parameters
   * @param params.key Feedback key for the evaluator
   * @param params.choices Array of valid categorical options
   * @param params.description Description of the scoring criteria
   * @param params.reasoningKey Optional key for the LLM reasoning/explanation for the score
   * @param params.reasoningDescription Optional description of score reasoning, provided to the LLM in the structured output
   */
  constructor(params: {
    key: string;
    choices: string[];
    description: string;
    reasoningKey?: string;
    reasoningDescription?: string;
  }) {
    this.key = params.key;
    this.choices = params.choices;
    this.description = params.description;
    this.reasoningKey = params.reasoningKey;
    this.reasoningDescription = params.reasoningDescription;
  }
}

/**
 * Configuration for continuous (numeric) scoring in evaluations.
 * Used to define scores that fall within a numeric range.
 */
export class ContinuousScoreConfig {
  /** Feedback key for the evaluator */
  key: string;
  /** Minimum allowed score value (defaults to 0) */
  min: number;
  /** Maximum allowed score value (defaults to 1) */
  max: number;
  /** Description of the scoring criteria */
  description: string;
  /** Optional key for the LLM reasoning/explanation for the score */
  reasoningKey?: string;
  /** Optional description of score reasoning, provided to the LLM in the structured output */
  reasoningDescription?: string;

  /**
   * Creates a new continuous score configuration
   * @param params Configuration parameters
   * @param params.key Feedback key for the evaluator
   * @param params.description Description of the scoring criteria
   * @param params.min Optional minimum score value (defaults to 0)
   * @param params.max Optional maximum score value (defaults to 1)
   * @param params.reasoningKey Optional key for the LLM reasoning/explanation for the score
   * @param params.reasoningDescription Optional description of score reasoning, provided to the LLM in the structured output
   */
  constructor(params: {
    key: string;
    description: string;
    min?: number;
    max?: number;
    reasoningKey?: string;
    reasoningDescription?: string;
  }) {
    this.key = params.key;
    this.min = params.min ?? 0;
    this.max = params.max ?? 1;
    this.description = params.description;
    this.reasoningKey = params.reasoningKey;
    this.reasoningDescription = params.reasoningDescription;
  }
}

type ScoreConfig = CategoricalScoreConfig | ContinuousScoreConfig;

function createScoreJsonSchema(scoreConfig: ScoreConfig): Record<string, any> {
  const properties: Record<string, any> = {};

  if (scoreConfig.reasoningKey) {
    properties[scoreConfig.reasoningKey] = {
      type: "string",
      description:
        scoreConfig.reasoningDescription ||
        "First, think step by step to explain your score.",
    };
  }

  if ("choices" in scoreConfig) {
    properties.value = {
      type: "string",
      enum: scoreConfig.choices,
      description: `The score for the evaluation, one of ${scoreConfig.choices.join(
        ", "
      )}.`,
    };
  } else {
    properties.score = {
      type: "number",
      minimum: scoreConfig.min,
      maximum: scoreConfig.max,
      description: `The score for the evaluation, between ${scoreConfig.min} and ${scoreConfig.max}, inclusive.`,
    };
  }

  return {
    title: scoreConfig.key,
    description: scoreConfig.description,
    type: "object",
    properties,
    required: scoreConfig.reasoningKey
      ? ["choices" in scoreConfig ? "value" : "score", scoreConfig.reasoningKey]
      : ["choices" in scoreConfig ? "value" : "score"],
  };
}

interface LLMEvaluatorParams {
  promptTemplate: string | [string, string][];
  scoreConfig: ScoreConfig;
  mapVariables?: (run: Run, example?: Example) => Record<string, any>;
  modelName?: string;
  modelProvider?: string;
  reasoningKey?: string;
}

export class LLMEvaluator implements RunEvaluator {
  prompt: any;
  mapVariables?: (run: Run, example?: Example) => Record<string, any>;
  scoreConfig: ScoreConfig;
  scoreSchema: Record<string, any>;
  runnable: any;
  reasoningKey: string;

  constructor() {}

  static async create(params: LLMEvaluatorParams): Promise<LLMEvaluator> {
    const evaluator = new LLMEvaluator();
    await evaluator.initialize(
      params.promptTemplate,
      params.scoreConfig,
      params.mapVariables,
      params.modelName || "gpt-4o",
      params.modelProvider || "openai"
    );
    return evaluator;
  }
  private async initialize(
    promptTemplate: string | [string, string][],
    scoreConfig: ScoreConfig,
    mapVariables?: (run: Run, example?: Example) => Record<string, any>,
    modelName: string = "gpt-4o",
    modelProvider: string = "openai"
  ) {
    try {
      // Store the configuration
      this.scoreConfig = scoreConfig;
      this.mapVariables = mapVariables;

      // Create the score schema
      this.scoreSchema = createScoreJsonSchema(scoreConfig);

      // Create the prompt template
      if (typeof promptTemplate === "string") {
        this.prompt = ChatPromptTemplate.fromMessages([
          { role: "human", content: promptTemplate },
        ]);
      } else {
        this.prompt = ChatPromptTemplate.fromMessages(promptTemplate);
      }

      // Initialize the chat model with structured output
      const chatModel = await initChatModel(modelName, {
        modelProvider: modelProvider,
      });

      // Create the runnable pipeline
      const modelWithStructuredOutput = chatModel.withStructuredOutput(
        this.scoreSchema
      );
      this.runnable = this.prompt.pipe(modelWithStructuredOutput);
    } catch (e) {
      throw new Error(
        "LLMEvaluator requires langchain to be installed. " +
          "Please install langchain by running `npm install langchain`."
      );
    }
  }

  async evaluateRun(
    run: Run,
    example?: Example
  ): Promise<EvaluationResult | EvaluationResults> {
    const runId = uuid.v4();
    const variables = this.prepareVariables(run, example);
    let output = await this.runnable.invoke(variables, { runId: runId });

    return this.parseOutput(output, runId);
  }

  private prepareVariables(run: Run, example?: Example): Record<string, any> {
    if (this.mapVariables) {
      return this.mapVariables(run, example);
    }

    const variables: Record<string, any> = {};

    // Input handling
    if (Object.keys(run.inputs).length === 0) {
      throw new Error(
        "No input keys are present in run.inputs but the prompt requires 'input'."
      );
    }
    if (Object.keys(run.inputs).length !== 1) {
      throw new Error(
        "Multiple input keys are present in run.inputs. Please provide a map_variables function."
      );
    }
    variables.input = Object.values(run.inputs)[0];

    // Output handling
    if (!run.outputs || Object.keys(run.outputs).length === 0) {
      throw new Error(
        "No output keys are present in run.outputs but the prompt requires 'output'."
      );
    }
    if (Object.keys(run.outputs).length !== 1) {
      throw new Error(
        "Multiple output keys are present in run.outputs. Please provide a map_variables function."
      );
    }
    variables.output = Object.values(run.outputs)[0];

    // Expected output handling
    if (example?.outputs) {
      if (Object.keys(example.outputs).length === 0) {
        throw new Error(
          "No output keys are present in example.outputs but the prompt requires 'expected'."
        );
      }
      if (Object.keys(example.outputs).length !== 1) {
        throw new Error(
          "Multiple output keys are present in example.outputs. Please provide a map_variables function."
        );
      }
      variables.expected = Object.values(example.outputs)[0];
    }

    return variables;
  }

  private parseOutput(
    output: Record<string, any>,
    runId: string
  ): EvaluationResult {
    if ("choices" in this.scoreConfig) {
      const value = output.value;
      const explanation = output[this.reasoningKey];
      return {
        key: this.scoreConfig.key,
        value,
        comment: explanation,
        sourceRunId: runId,
      };
    } else {
      const score = output.score;
      const explanation = output[this.reasoningKey];
      return {
        key: this.scoreConfig.key,
        score,
        comment: explanation,
        sourceRunId: runId,
      };
    }
  }
}
